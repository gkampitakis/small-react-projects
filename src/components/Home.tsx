import React, { ReactElement } from 'react';
import routes from '../routes';
import { Link } from 'react-router-dom';


export default function Home (): ReactElement {
  return (
    <main>
      <header >
        Small React Projects
      </header>
      <section>
        {routes.map(({ projectName, route }, idx) => (
          <Link key={idx} to={route}>
            {projectName}
          </Link>
        ))}
      </section>
      <footer>
        G Kampitakis
      </footer>
    </main>
  );
}
