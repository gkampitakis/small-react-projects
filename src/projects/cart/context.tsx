import React, { useContext, useReducer, useEffect, createContext, ReactElement, ReducerAction } from 'react';

const URL = 'https://course-api.com/react-useReducer-cart-project';
const AppContext = createContext<DispatchActions & CartState>({} as any);

interface DispatchActions {
  clearCart: () => void;
  increase: (id: number) => void;
  decrease: (id: number) => void;
  remove: (id: number) => void;
}

interface Cart {
  id: number;
  img: string;
  title: string;
  price: number;
  amount: number;
}

interface CartState {
  cart: Cart[];
  loading: boolean;
  total: number;
  amount: number;
}

type Action =
  | { type: 'DISPLAY_ITEMS', payload: { data: Cart[] } }
  | { type: 'REMOVE', payload: { id: number } }
  | { type: 'INCREASE', payload: { id: number } }
  | { type: 'DECREASE', payload: { id: number } }
  | { type: 'GET_TOTALS' }
  | { type: 'CLEAR_CART' }
  | { type: 'LOADING' };

const AppProvider = ({ children }: { children: ReactElement | ReactElement[] }): ReactElement => {
  const [state, dispatch] = useReducer(reducer, {
    cart: [],
    loading: false,
    total: 0,
    amount: 0
  });

  async function fetchData () {
    dispatch({ type: 'LOADING' });

    console.log('fetching data');


    const response = await fetch(URL);
    const data = await response.json();

    dispatch({ type: 'DISPLAY_ITEMS', payload: { data } });
  }

  function decrease (id: number) {
    dispatch({ type: 'DECREASE', payload: { id } });
  }

  function increase (id: number) {
    dispatch({ type: 'INCREASE', payload: { id } });
  }

  function clearCart () {
    dispatch({ type: 'CLEAR_CART' });
  }

  function remove (id: number) {
    dispatch({ type: 'REMOVE', payload: { id } });
  }

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    dispatch({ type: 'GET_TOTALS' });
  }, [state.cart]);

  return (
    <AppContext.Provider value={{
      ...state,
      clearCart,
      increase,
      decrease,
      remove
    }}>
      {children}
    </AppContext.Provider>
  )
};

export const useGlobalContext = (): DispatchActions & CartState => useContext(AppContext);

export { AppContext, AppProvider };

function reducer (state: CartState, action: Action): CartState {
  switch (action.type) {
    case 'INCREASE':
      return {
        ...state,
        cart: state.cart.map((item) => {
          if (item.id === action.payload.id) return {
            ...item,
            amount: item.amount + 1
          };
          return item;
        })
      }
    case 'DECREASE':
      return {
        ...state,
        cart: state.cart.map((item) => {
          if (item.id === action.payload.id) return {
            ...item,
            amount: item.amount - 1
          };
          return item;
        }).filter(i => !!i.amount)
      }
    case 'LOADING':
      return {
        ...state,
        loading: true
      };
    case 'DISPLAY_ITEMS':
      return {
        ...state,
        cart: action.payload.data,
        loading: false
      };
    case 'GET_TOTALS':
      return {
        ...state,
        ...state.cart.reduce((acc, item) => {
          const { amount, price } = item;
          const total = amount * price;

          return { amount: acc.amount + amount, total: acc.total + total }
        }, { amount: 0, total: 0 })
      }
    case 'REMOVE':
      return {
        ...state,
        cart: state.cart.filter(i => i.id !== action.payload.id)
      }
    case 'CLEAR_CART':
      return {
        ...state,
        cart: []
      }
    default:
      throw Error('Reached to unknown state in reducer');
  }
}