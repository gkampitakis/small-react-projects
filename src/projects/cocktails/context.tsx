import React, {
  createContext,
  ReactElement,
  useContext,
  useState,
  useEffect,
  useRef
} from 'react';
import { useFetch } from '../../hooks';

const URL = 'https://www.thecocktaildb.com/api/json/v1/1/search.php?s=';
const DEBOUNCE_TIMER = 2000;
const AppContext = createContext<{
  searchCocktail: (cocktail: string) => void;
  refetch: () => void;
  loading: boolean;
  bounceLoading: boolean;
  error: string | null;
  cocktailList: CocktailI[];
}>({} as any);

export interface CocktailI {
  id: string;
  name: string;
  info: string;
  glass: string;
  image: string;
}

const AppProvider = ({
  children
}: {
  children: ReactElement | ReactElement[];
}): ReactElement => {
  const [url, setUrl] = useState(`${URL}a`);
  const [bounceLoading, setBounceLoading] = useState(false);
  const [data, loading, , error] = useFetch<any>(url);
  const [cocktailList, setCocktailList] = useState<CocktailI[]>([]);
  const searchDebounce = useRef<ReturnType<typeof setTimeout> | undefined>(
    undefined
  );

  function searchCocktail(cocktail: string) {
    setBounceLoading(true);
    if (searchDebounce.current) clearTimeout(searchDebounce.current);

    searchDebounce.current = setTimeout(() => {
      console.info('Sending request for ' + cocktail);
      setBounceLoading(false);
      setUrl(`${URL}${cocktail}`);
    }, DEBOUNCE_TIMER);
  }

  function filterResults(data: any[]): CocktailI[] {
    if (!data) return [];

    return data.map(
      ({ idDrink, strDrink, strDrinkThumb, strAlcoholic, strGlass }) => ({
        id: idDrink,
        name: strDrink,
        image: strDrinkThumb,
        info: strAlcoholic,
        glass: strGlass
      })
    );
  }

  function refetch() {
    setUrl(`${URL}a`);
  }

  useEffect(() => {
    setCocktailList(filterResults(data?.drinks));
  }, [data]);

  return (
    <AppContext.Provider
      value={{
        searchCocktail,
        refetch,
        loading,
        error,
        cocktailList,
        bounceLoading
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useGlobalContext = () => useContext(AppContext);
export { AppProvider };
