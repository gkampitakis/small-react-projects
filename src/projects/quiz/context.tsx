import React, {
  createContext,
  ReactElement,
  useContext,
  useState,
  ChangeEvent
} from 'react';

interface QuizSettings {
  amount: number;
  category: 'sports' | 'history' | 'politics';
  difficulty: 'easy' | 'medium' | 'hard';
}

type Question = {
  category: string;
  type: string;
  difficulty: string;
  question: string;
  correct_answer: string;
  incorrect_answers: string[];
};

enum CategoryCodes {
  sports = 21,
  history,
  politics
}

const API_ENDPOINT = 'https://opentdb.com/api.php?';
const AppContext = createContext<{
  waiting: boolean;
  error: Error | null;
  loading: boolean;
  quiz: QuizSettings;
  questions: Question[];
  isModalOpen: boolean;
  currentQ: number;
  correct: number;
  openModal: () => void;
  checkAnswer: (value: any) => void;
  nextQuestion: () => void;
  closeModal: () => void;
  setupFormOnChange: (
    e: ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => void;
  setupFormOnSubmit: (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
}>({} as any);

const AppProvider = ({
  children
}: {
  children: ReactElement | ReactElement[];
}): ReactElement => {
  const [waiting, setWaiting] = useState(true);
  const [loading, setLoading] = useState(false);
  const [questions, setQuestions] = useState<Question[]>([]);
  const [error, setError] = useState(null);
  const [currentQ, setCurrentQ] = useState(0);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [correct, setCorrect] = useState(0);
  const [quiz, setQuiz] = useState<QuizSettings>({
    amount: 10,
    category: 'sports',
    difficulty: 'easy'
  });

  function setupFormOnChange(
    e: ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) {
    const { value, name } = e.target;

    setQuiz({ ...quiz, [name]: value });
  }

  function openModal() {
    setIsModalOpen(true);
  }

  function closeModal() {
    setWaiting(true);
    setCorrect(0);
    setIsModalOpen(false);
  }

  function checkAnswer(value: boolean) {
    if (value) {
      setCorrect(oldState => oldState + 1);
    }
    nextQuestion();
  }

  function nextQuestion() {
    setCurrentQ(i => {
      const index = i + 1;
      if (index > questions.length - 1) {
        console.log('open');

        openModal();
        return 0;
      } else {
        return index;
      }
    });
  }

  function setupFormOnSubmit(
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) {
    e.preventDefault();

    const { amount, category, difficulty } = quiz;
    const url = `${API_ENDPOINT}amount=${amount}&difficulty=${difficulty}&category=${CategoryCodes[category]}&type=multiple`;

    setLoading(true);
    setWaiting(false);

    fetch(url)
      .then(d => d.json())
      .then(d => {
        setQuestions(d.results);
        console.log(d.results);
      })
      .catch(e => {
        setError(e);
        setWaiting(true);
      })
      .finally(() => setLoading(false));
  }

  return (
    <AppContext.Provider
      value={{
        waiting,
        error,
        loading,
        quiz,
        questions,
        isModalOpen,
        currentQ,
        correct,
        nextQuestion,
        checkAnswer,
        closeModal,
        openModal,
        setupFormOnChange,
        setupFormOnSubmit
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

const useGlobalContext = () => useContext(AppContext);

export { useGlobalContext, AppProvider };
