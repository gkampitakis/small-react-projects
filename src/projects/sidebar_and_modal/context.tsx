import React, { useState, useContext, ReactElement } from 'react';


const AppContext = React.createContext<{
  isSidebarOpen: boolean;
  isModalOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
  openSidebar: () => void;
  closeSidebar: () => void;
}>({} as any);


const AppProvider = ({ children }: { children: ReactElement | ReactElement[] }): ReactElement => {
  const [isSidebarOpen, setIsSidebarOpen] = useState(false);
  const [isModalOpen, setIsModalOpen] = useState(false);

  const openModal = () => setIsModalOpen(true);
  const closeModal = () => setIsModalOpen(false);
  const openSidebar = () => setIsSidebarOpen(true);
  const closeSidebar = () => setIsSidebarOpen(false);

  return (
    < AppContext.Provider value={{
      isSidebarOpen,
      isModalOpen,
      openModal,
      closeModal,
      openSidebar,
      closeSidebar
    }}>
      {children}
    </AppContext.Provider >
  );
}

export const useGlobalContext = () => useContext(AppContext);

export { AppProvider };
