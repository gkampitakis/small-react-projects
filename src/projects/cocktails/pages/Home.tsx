import React, { ReactElement } from 'react';
import { useGlobalContext } from '../context';
import Search from '../components/Search';
import List from '../components/List';
import ErrorCmp from '../../../components/Error';
import Loading from '../../../components/Loading';

export default function Home(): ReactElement {
  const { loading, cocktailList, error } = useGlobalContext();
  const show = loading && !cocktailList.length;

  return (
    <>
      <Search />
      {!show && !error && <List items={cocktailList} />}
      {show && <Loading />}
      {error && <ErrorCmp message={error} />}
    </>
  );
}
