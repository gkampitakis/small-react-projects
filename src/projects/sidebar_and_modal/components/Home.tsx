import React, { ReactElement } from 'react';
import { FaBars } from 'react-icons/fa';
import { useGlobalContext } from '../context';


export default function Home (): ReactElement {
  const { openSidebar, openModal } = useGlobalContext();

  return (
    <main>
      <button onClick={openSidebar} className="sidebar-toggle" data-testid='open-sidebar'>
        <FaBars />
      </button>
      <button onClick={openModal} className="show-btn">
        show modal
      </button>
    </main>
  );
}
