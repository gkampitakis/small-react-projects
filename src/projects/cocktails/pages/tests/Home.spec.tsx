import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Home from '../Home';


jest.mock('../../context', () => ({
  useGlobalContext: jest.fn()
}));

import { useGlobalContext } from '../../context';
import { MemoryRouter } from 'react-router-dom';
const useGlobalContextMock = useGlobalContext as jest.Mock;

describe('Home', () => {

  beforeEach(jest.resetAllMocks);

  it('Should render correctly', () => {
    useGlobalContextMock.mockImplementation(() => ({
      loading: false,
      error: false,
      cocktailList: []
    }));

    const tree = renderer
      .create(<Home />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render loading prompt', () => {
    useGlobalContextMock.mockImplementation(() => ({
      loading: true,
      error: false,
      cocktailList: []
    }));

    const { baseElement } = render(<Home />);

    expect(baseElement).toMatchSnapshot();
  });

  it('Should not render loading prompt', () => {
    useGlobalContextMock.mockImplementation(() => ({
      loading: true,
      error: false,
      cocktailList: [{
        id: 0,
        name: 'mockCocktail',
        info: 'mockInfo',
        glass: 'mockGlass',
        image: 'mockImg'
      }]
    }));
    const { baseElement } = render(<Home />, { wrapper: MemoryRouter });

    expect(baseElement).toMatchSnapshot();
  });

  it('Should render error prompt', () => {
    useGlobalContextMock.mockImplementation(() => ({
      loading: false,
      error: 'MockError',
      cocktailList: []
    }));

    const { baseElement } = render(<Home />, { wrapper: MemoryRouter });

    expect(baseElement).toMatchSnapshot();
  });
});
