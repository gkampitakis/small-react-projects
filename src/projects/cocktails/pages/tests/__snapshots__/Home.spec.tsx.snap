// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should not render loading prompt 1`] = `
<body>
  <div>
    <section
      class="search_bar"
    >
      <form>
        <label
          for="search"
        >
          Search your favorite cocktail
        </label>
        <div>
          <input
            id="search"
            name="search"
            placeholder="e.g. Sex on the beach"
            type="text"
            value=""
          />
          
        </div>
        <div
          class="loader undefined"
        >
          <div
            class="bar"
          />
        </div>
      </form>
    </section>
    <section
      class="cocktail-list"
    >
      <h2
        class="title"
      >
        Cocktails
      </h2>
      <div
        class="cocktails-center"
      >
        <article
          class="cocktail"
        >
          <div
            class="img-container"
          >
            <img
              alt="mockCocktail"
              src="mockImg"
            />
          </div>
          <div
            class="cocktail-footer"
          >
            <h3>
              mockCocktail
            </h3>
            <h4>
              mockGlass
            </h4>
            <p>
              mockInfo
            </p>
            <a
              class="btn"
              href="/cocktails/details/0"
            >
              details
            </a>
          </div>
        </article>
      </div>
    </section>
  </div>
</body>
`;

exports[`Home Should render correctly 1`] = `
Array [
  <section
    className="search_bar"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="search"
      >
        Search your favorite cocktail
      </label>
      <div>
        <input
          id="search"
          name="search"
          onChange={[Function]}
          placeholder="e.g. Sex on the beach"
          type="text"
          value=""
        />
        
      </div>
      <div
        className="loader undefined"
      >
        <div
          className="bar"
        />
      </div>
    </form>
  </section>,
  <h2
    className="title"
  >
    no cocktails match your search
  </h2>,
]
`;

exports[`Home Should render error prompt 1`] = `
<body>
  <div>
    <section
      class="search_bar"
    >
      <form>
        <label
          for="search"
        >
          Search your favorite cocktail
        </label>
        <div>
          <input
            id="search"
            name="search"
            placeholder="e.g. Sex on the beach"
            type="text"
            value=""
          />
          
        </div>
        <div
          class="loader undefined"
        >
          <div
            class="bar"
          />
        </div>
      </form>
    </section>
    <main
      class="error_component"
    >
      <img
        alt="Error page image"
        src="error.svg"
      />
      <footer>
        <h2>
          MockError
        </h2>
      </footer>
    </main>
  </div>
</body>
`;

exports[`Home Should render loading prompt 1`] = `
<body>
  <div>
    <section
      class="search_bar"
    >
      <form>
        <label
          for="search"
        >
          Search your favorite cocktail
        </label>
        <div>
          <input
            id="search"
            name="search"
            placeholder="e.g. Sex on the beach"
            type="text"
            value=""
          />
          
        </div>
        <div
          class="loader undefined"
        >
          <div
            class="bar"
          />
        </div>
      </form>
    </section>
    <main
      class="loading_component"
    >
      <div
        class="loader"
      >
        <div
          class="bar"
        />
      </div>
      <img
        alt="Loading Logo"
        class="logo"
        src="loading.png"
      />
    </main>
  </div>
</body>
`;
